### PR TITLE
add scorers for token classification

### LINKED ISSUE #1869: [FEAT] Add support for token classification metrics
Research which metrics are common. Can we use also ranking metrics using sklearn?

Add support in them to the DeepChecksScorer

No need to support them in tabular checks

See if easy to extract relevant stuff from tabular metrics to core (because used by all packages no only tabular)

### PR DESCRIPTION
**Pull request by noamzbr**
#### Reference Issues/PRs

Resolves #1869

### CODE REVIEW

File: deepchecks/nlp/text_data.py [hunk 1]
```diff
@@ -318,15 +318,21 @@ def classes(self) -> t.Optional[t.List[t.Union[str, int]]]:
...
+                label_set = self.get_tokens(self._label)
             else:
                 raise DeepchecksValueError(f'Task type {self.task_type} is not supported.')
             self._classes = sorted(list(label_set))
         return self._classes
 
+    @staticmethod
+    def get_tokens(token_annotations: t.Sequence[t.Sequence[t.Tuple[str, int, int, t.Any]]]) -> t.Set[str]:
+        """Return the token strings from token classification labels or predictions."""
+        return set().union(*[
```
Comment (Nadav-Barak) [inline deepchecks/nlp/text_data.py:330]: this is hard to read, please use regular for loop instead of list comprehension for this

File: tests/nlp/checks/model_evaluation/single_dataset_performance_nlp_test.py [hunk 2]
```diff
@@ -105,3 +105,46 @@ def test_run_with_scorer_multilabel_class_names(text_multilabel_classification_d
     # Assert
     assert_that(result.value.values[0][-1], close_to(1.0, 0.001))
     assert_that(result.value.values[0][0], equal_to('a'))
+
+
+def test_run_with_scorer_token(text_token_classification_dataset_mock):
```
Comment (Nadav-Barak) [inline tests/nlp/checks/model_evaluation/single_dataset_performance_nlp_test.py:110]: Also check different scenarios of wrong input format,
order of stuff in tuples in predictions
order of stuff in tuples in labels
labels and prediction that disagree - in labels str, in locations

File: deepchecks/nlp/checks/model_evaluation/single_dataset_performance.py [hunk 3]
```diff
@@ -46,12 +49,15 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
         """Run check."""
         dataset = context.get_data_by_kind(dataset_kind)
         model = context.model
-        scorers = context.get_scorers(self.scorers, use_avg_defaults=False)
+        span_aligner = SpanAligner()
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/model_evaluation/single_dataset_performance.py:52]: I think this should be created and employing in a more general manner, is this used only for performance checks?
Consider placing it in context (usecase in mind in for evaluation suite)

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+__all__ = ['get_default_token_scorers']
+
+
+DEFAULT_AVG_SCORER_NAMES = ('f1_macro', 'recall_macro', 'precision_macro')
+DEFAULT_PER_CLASS_SCORER_NAMES = ('f1_per_class', 'f1_per_class', 'f1_per_class')
+
+
+class SpanAligner:
+    """An object for transforming the labels and predictions of a specific dataset"""
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:32]: into what?

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+    def fit(self, y_true: TTokenLabel, y_pred: 'TTokenPred'):
+        """Fit the SpanAligner object to the labels and predictions of a specific dataset."""
+        self.processed_labels, self.processed_predictions = self._spans_to_token_list(y_true, y_pred)
+
+        # Post process class list to match the class order received by seqeval metrics
+        self.classes.discard('O')
+        self.classes = sorted(list(self.classes))
+
+    def fit_transform(self, y_true: TTokenLabel, y_pred: 'TTokenPred'
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:69]: Also add a transform method and have the fit_transform use it

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            predictions are considered if they overlap any label span.
+        proba_threshold: float, default: 0
+            The minimal probability score for a prediction to be considered.
+        """
+        if not isinstance(strict_alignment, bool):
+            raise TypeError('strict_alignment must be a boolean')
+        if not strict_alignment:
+            # TODO: Implement non-strict alignment, probably using https://github.com/biocore-ntnu/ncls or similar
+            raise NotImplementedError('Non Strict alignment is not yet implemented')
+        if not isinstance(proba_threshold, float):
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:50]: merge 2 if on proba_threshold into one line

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        self.classes = sorted(list(self.classes))
+
+    def fit_transform(self, y_true: TTokenLabel, y_pred: 'TTokenPred'
+                      ) -> t.Tuple[t.List[t.List[str]], t.List[t.List[str]]]:
+        """Fit (if necessary) and transform the labels and predictions of a specific dataset."""
+        if self.processed_labels is None or self.processed_predictions is None:
+            self.fit(y_true, y_pred)
+        return self.processed_labels, self.processed_predictions
+
+    def _spans_to_token_list(self, y_true: TTokenLabel, y_pred: 'TTokenPred',
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:76]: no TTokenPred type?

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                             ) -> t.Tuple[t.List[t.List[str]], t.List[t.List[str]]]:
+        """Convert from predictions and labels in span format to a list of token annotations.
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:78]: Provide info here or in init what the format should look like

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+                                               list(map(lambda x: x[0], label_spans)), side='left')
+                # Elements of sample_y_pred that match their counterparts in label_spans
+                pred_index = np.take(np.arange(len(pred_spans)), sorted_index, mode='clip')
+                bool_matching = np.array(list(map(hash, pred_spans)))[pred_index] == np.array(list(map(hash, label_spans)))
+                # Filter out predictions that do not match their corresponding labels or their threshold is too low.
+                # Predictions that where filtered are filled with 'O'.
+                sample_predictions = \
+                    [sample_y_pred[pred_index[idx]][0] if
+                     ((sample_y_pred[pred_index[idx]][3] >= self.proba_threshold) and bool_matching[idx]) else 'O'
+                     for idx in range(len(sample_y_true))]
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:118]: This is hard to understand, so I cant properly review.
easy ideas to improve:
Better variable names, explain at start what is the expected format to be in sample_prediction

File: deepchecks/nlp/context.py [hunk 10]
```diff
@@ -375,7 +375,7 @@ def get_scorers(self,
             else:
                 scorers = scorers or get_default_scorers(TabularTaskType.BINARY, use_avg_defaults)
         elif self.task_type == TaskType.TOKEN_CLASSIFICATION:
-            scorers = []  # TODO: Complete that
+            scorers = get_default_token_scorers(scorers, span_aligner, use_avg_defaults)
```
Comment (Nadav-Barak) [inline deepchecks/nlp/context.py:378]: Create two different paths here, only use get_default_token_scorers if no scorers are provided.
If scorers are provided process them here or create a different function for that
Currently get_default_token_scorers is hard function to follow

**Review by Nadav-Barak — COMMENTED** [review]
Will need to rereview the scorer related stuff

File: deepchecks/nlp/checks/model_evaluation/single_dataset_performance.py (hunk 3 above)
Comment (noamzbr) [reply deepchecks/nlp/checks/model_evaluation/single_dataset_performance.py:52]: This can be done only assuming all check run on the exact same sample. It seems reasonable so I'm willing to do it, but then we need to remember not to enable different samples for different checks

File: deepchecks/nlp/metric_utils/token_classification.py (hunk 7 above)
Comment (noamzbr) [reply deepchecks/nlp/metric_utils/token_classification.py:76]: Can't import it due to circular import (defined in context, context uses this file)

File: deepchecks/nlp/metric_utils/token_classification.py (hunk 5 above)
Comment (noamzbr) [reply deepchecks/nlp/metric_utils/token_classification.py:69]: Implemented differently (called it get_transformed so that it's clear it does not do the transformation, just return cached)

File: deepchecks/nlp/checks/model_evaluation/single_dataset_performance.py (hunk 3 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/checks/model_evaluation/single_dataset_performance.py:52]: not necessarily, you can re-initialize it if needed

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+__all__ = ['get_default_token_scorers', 'validate_scorers', 'get_scorer_dict']
+
+DEFAULT_AVG_SCORER_NAMES = ('f1_macro', 'recall_macro', 'precision_macro')
+DEFAULT_PER_CLASS_SCORER_NAMES = ('f1_per_class', 'f1_per_class', 'f1_per_class')
+
+
+class SpanAligner:
+    """An object for transforming the labels and predictions of a specific dataset.
+
+    Labels come in the deepchecks accepted format and are returned as a list of lists of string annotations.
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:32]: Can add more detail how the output format should look like. maybe add an example

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+        Parameters:
+        -----------
+            y_true: TTokenLabel
+                Labels as provided by the user in the accepted deepchecks format (TTokenLabel, detailed in TextData)
+            y_pred: TTokenPred
+                Predictions as provided by the user in the accepted format (TTokenPred, detailed in _shared_docs.py)
+
+        Returns:
+        --------
+            A tuple of processed labels and processed predictions, each a list of lists of string annotations.
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:105]: same here

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+            raise DeepchecksValueError(f'Must have both a label entry and a prediction entry for each sample, but '
+                                       f'found {len(y_true)} label entries and {len(y_pred)} prediction entries')
+
+        labels = []
+        predictions = []
+        for sample_idx in range(len(y_true)):
+            # Sort label and predictions by the starting position of the span
+            sample_y_true = sorted(y_true[sample_idx], key=lambda x: x[1])
+            sample_y_pred = sorted(y_pred[sample_idx], key=lambda x: x[1])
+            sample_labels = [annotation[0] for annotation in sample_y_true]
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:117]: This variable is redundant (can move line 149 out of the loop and use labels)

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+            labels.append(sample_labels)
+
+            # A list of the spans, sorted by their starting position
+            label_spans = [(annotation[1], annotation[2]) for annotation in sample_y_true]
+            pred_spans = [(annotation[1], annotation[2]) for annotation in sample_y_pred]
+
+            # Here we want to find the prediction spans that match specific label spans, and discard to rest.
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:124]:
```suggestion
            # Here we want to find the prediction spans that match specific label spans, and discard the rest.
```

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+
+
+def make_token_scorer(metric: t.Callable[[t.List[t.List[str]], t.List[t.List[str]]], float],
+                      span_aligner: SpanAligner,
+                      **kwargs):
+    """Make a scorer that accepts span labels and predictions."""
+    return make_scorer(make_modified_metric(metric, span_aligner, **kwargs))
+
+
+def get_scorer_dict(span_aligner: SpanAligner,
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:175]: Move this to start on class

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+
+def validate_scorers(scorers: t.List[str], span_aligner: SpanAligner):
+    """Validate the given scorer list"""
+    scoring_dict = get_scorer_dict(span_aligner)
+
+    if not isinstance(scorers, Sequence):
+        raise DeepchecksValueError(f'Scorers must be a Sequence, got {type(scorers)}')
+    if not all(isinstance(name, str) for name in scorers):
```
Comment (Nadav-Barak) [inline deepchecks/nlp/metric_utils/token_classification.py:220]: Add TODO here regarding future support for custom scorers

File: tests/nlp/checks/model_evaluation/single_dataset_performance_nlp_test.py (hunk 2 above)
Comment (Nadav-Barak) [reply tests/nlp/checks/model_evaluation/single_dataset_performance_nlp_test.py:110]: @noamzbr

**Review by Nadav-Barak — APPROVED** [review]
See requests for test :)

File: tests/nlp/checks/model_evaluation/single_dataset_performance_nlp_test.py (hunk 2 above)
Comment (noamzbr) [reply tests/nlp/checks/model_evaluation/single_dataset_performance_nlp_test.py:110]: labels and prediction that disagree in str and locations is tested in test_run_with_scorer_token_per_class, can be seen in test_run_with_scorer_token_per_class.
Will implement the rest
